File: src/core/schema_tests/python/test_controller.py
# SPDX-FileCopyrightText: Copyright (c) 2025-2026 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
# SPDX-License-Identifier: Apache-2.0

"""Unit tests for Controller types in isaacteleop.schema.

Tests the following FlatBuffers types:
- ControllerInputState: Struct with button and axis inputs (immutable)
- ControllerPose: Struct with pose and validity (immutable)
- ControllerSnapshot: Table representing complete controller state
"""

import pytest

from isaacteleop.schema import (
    ControllerInputState,
    ControllerPose,
    ControllerSnapshot,
    ControllerSnapshotRecord,
    Pose,
    Point,
    Quaternion,
    DeviceDataTimestamp,
)


class TestControllerInputState:
    """Tests for ControllerInputState struct (immutable)."""

    def test_default_construction(self):
        """Test default construction creates ControllerInputState with default values."""
        inputs = ControllerInputState()

        assert inputs is not None
        assert inputs.primary_click is False
        assert inputs.secondary_click is False
        assert inputs.thumbstick_click is False
        assert inputs.menu_click is False
        assert inputs.thumbstick_x == pytest.approx(0.0)
        assert inputs.thumbstick_y == pytest.approx(0.0)
        assert inputs.squeeze_value == pytest.approx(0.0)
        assert inputs.trigger_value == pytest.approx(0.0)

    def test_button_states(self):
        """Test constructing with button states."""
        inputs = ControllerInputState(
            primary_click=True,
            secondary_click=True,
            thumbstick_click=True,
            menu_click=True,
            thumbstick_x=0.0,
            thumbstick_y=0.0,
            squeeze_value=0.0,
            trigger_value=0.0,
        )

        assert inputs.primary_click is True
        assert inputs.secondary_click is True
        assert inputs.thumbstick_click is True
        assert inputs.menu_click is True

    def test_analog_values(self):
        """Test constructing with analog axis values."""
        inputs = ControllerInputState(
            primary_click=False,
            secondary_click=False,
            thumbstick_click=False,
            menu_click=False,
            thumbstick_x=0.5,
            thumbstick_y=-0.75,
            squeeze_value=0.8,
            trigger_value=1.0,
        )

        assert inputs.thumbstick_x == pytest.approx(0.5)
        assert inputs.thumbstick_y == pytest.approx(-0.75)
        assert inputs.squeeze_value == pytest.approx(0.8)
        assert inputs.trigger_value == pytest.approx(1.0)

    def test_repr(self):
        """Test __repr__ method."""
        inputs = ControllerInputState(
            primary_click=True,
            secondary_click=False,
            thumbstick_click=False,
            menu_click=False,
            thumbstick_x=0.0,
            thumbstick_y=0.0,
            squeeze_value=0.0,
            trigger_value=0.5,
        )

        repr_str = repr(inputs)
        assert "ControllerInputState" in repr_str
        assert "primary=True" in repr_str


class TestControllerPose:
    """Tests for ControllerPose struct (immutable)."""

    def test_default_construction(self):
        """Test default construction creates ControllerPose with default values."""
        controller_pose = ControllerPose()

        assert controller_pose is not None
        # Structs always have default values, never None
        assert controller_pose.pose is not None
        assert controller_pose.is_valid is False

    def test_construction_with_pose(self):
        """Test constructing with pose data."""
        position = Point(1.0, 2.0, 3.0)
        orientation = Quaternion(0.0, 0.0, 0.0, 1.0)
        pose = Pose(position, orientation)
        controller_pose = ControllerPose(pose, True)

        assert controller_pose.pose.position.x == pytest.approx(1.0)
        assert controller_pose.pose.position.y == pytest.approx(2.0)
        assert controller_pose.pose.position.z == pytest.approx(3.0)
        assert controller_pose.pose.orientation.w == pytest.approx(1.0)
        assert controller_pose.is_valid is True

    def test_is_valid_flag(self):
        """Test is_valid flag."""
        position = Point(0.0, 0.0, 0.0)
        orientation = Quaternion(0.0, 0.0, 0.0, 1.0)
        pose = Pose(position, orientation)
        controller_pose = ControllerPose(pose, True)

        assert controller_pose.is_valid is True

    def test_pose_with_rotation(self):
        """Test pose with a rotation quaternion."""
        position = Point(0.0, 0.0, 0.0)
        # 90-degree rotation around Z axis
        orientation = Quaternion(0.0, 0.0, 0.7071068, 0.7071068)
        pose = Pose(position, orientation)
        controller_pose = ControllerPose(pose, True)

        assert controller_pose.pose.orientation.z == pytest.approx(0.7071068, abs=1e-5)
        assert controller_pose.pose.orientation.w == pytest.approx(0.7071068, abs=1e-5)
        assert controller_pose.is_valid is True

    def test_repr_without_valid_pose(self):
        """Test __repr__ when pose is not valid."""
        controller_pose = ControllerPose()
        repr_str = repr(controller_pose)

        assert "ControllerPose" in repr_str
        assert "is_valid=False" in repr_str

    def test_repr_with_pose(self):
        """Test __repr__ when pose is set."""
        position = Point(1.0, 2.0, 3.0)
        orientation = Quaternion(0.0, 0.0, 0.0, 1.0)
        pose = Pose(position, orientation)
        controller_pose = ControllerPose(pose, True)

        repr_str = repr(controller_pose)
        assert "ControllerPose" in repr_str
        assert "is_valid=True" in repr_str


class TestControllerSnapshot:
    """Tests for ControllerSnapshot table."""

    def test_default_construction(self):
        """Test default construction creates ControllerSnapshot with default values."""
        snapshot = ControllerSnapshot()

        assert snapshot is not None
        assert snapshot.grip_pose is not None
        assert snapshot.aim_pose is not None
        assert snapshot.inputs is not None

    def test_complete_snapshot(self):
        """Test creating a complete controller snapshot with all fields."""
        # Create poses
        grip_position = Point(0.1, 0.2, 0.3)
        grip_orientation = Quaternion(0.0, 0.0, 0.0, 1.0)
        grip_pose_data = Pose(grip_position, grip_orientation)
        grip_pose = ControllerPose(grip_pose_data, True)

        aim_position = Point(0.15, 0.25, 0.35)
        aim_orientation = Quaternion(0.0, 0.1, 0.0, 0.99)
        aim_pose_data = Pose(aim_position, aim_orientation)
        aim_pose = ControllerPose(aim_pose_data, True)

        # Create inputs
        inputs = ControllerInputState(
            primary_click=True,
            secondary_click=False,
            thumbstick_click=True,
            menu_click=False,
            thumbstick_x=0.5,
            thumbstick_y=-0.5,
            squeeze_value=0.75,
            trigger_value=1.0,
        )

        # Create snapshot
        snapshot = ControllerSnapshot(grip_pose, aim_pose, inputs)

        # Verify all fields
        assert snapshot.grip_pose.is_valid is True
        assert snapshot.grip_pose.pose.position.x == pytest.approx(0.1)
        assert snapshot.aim_pose.is_valid is True
        assert snapshot.aim_pose.pose.position.x == pytest.approx(0.15)
        assert snapshot.inputs.primary_click is True
        assert snapshot.inputs.trigger_value == pytest.approx(1.0)

    def test_repr_with_default(self):
        """Test __repr__ with default values."""
        snapshot = ControllerSnapshot()
        repr_str = repr(snapshot)

        assert "ControllerSnapshot" in repr_str


class TestControllerIntegration:
    """Integration tests combining multiple controller types."""

    def test_left_and_right_different_states(self):
        """Test that left and right controllers can have different states."""
        # Create left controller (active)
        left_position = Point(-0.2, 0.0, 0.0)
        left_orientation = Quaternion(0.0, 0.0, 0.0, 1.0)
        left_pose_data = Pose(left_position, left_orientation)
        left_grip = ControllerPose(left_pose_data, True)
        left_aim = ControllerPose(left_pose_data, True)
        left_inputs = ControllerInputState(
            primary_click=True,
            secondary_click=False,
            thumbstick_click=False,
            menu_click=False,
            thumbstick_x=0.0,
            thumbstick_y=0.0,
            squeeze_value=0.0,
            trigger_value=0.5,
        )
        left_snapshot = ControllerSnapshot(left_grip, left_aim, left_inputs)

        # Create right controller (default)
        right_snapshot = ControllerSnapshot()

        # Verify different states
        assert left_snapshot.inputs.trigger_value == pytest.approx(0.5)
        assert right_snapshot.inputs.trigger_value == pytest.approx(0.0)

    def test_default_controller(self):
        """Test representing a default-constructed controller."""
        snapshot = ControllerSnapshot()

        assert snapshot.grip_pose.is_valid is False
        assert snapshot.aim_pose.is_valid is False


class TestControllerEdgeCases:
    """Edge case tests for controller types."""

    def test_zero_analog_values(self):
        """Test with all analog values at zero (centered/released)."""
        inputs = ControllerInputState(
            primary_click=False,
            secondary_click=False,
            thumbstick_click=False,
            menu_click=False,
            thumbstick_x=0.0,
            thumbstick_y=0.0,
            squeeze_value=0.0,
            trigger_value=0.0,
        )

        assert inputs.thumbstick_x == pytest.approx(0.0)
        assert inputs.thumbstick_y == pytest.approx(0.0)
        assert inputs.squeeze_value == pytest.approx(0.0)
        assert inputs.trigger_value == pytest.approx(0.0)

    def test_max_analog_values(self):
        """Test with all analog values at maximum."""
        inputs = ControllerInputState(
            primary_click=False,
            secondary_click=False,
            thumbstick_click=False,
            menu_click=False,
            thumbstick_x=1.0,
            thumbstick_y=1.0,
            squeeze_value=1.0,
            trigger_value=1.0,
        )

        assert inputs.thumbstick_x == pytest.approx(1.0)
        assert inputs.thumbstick_y == pytest.approx(1.0)
        assert inputs.squeeze_value == pytest.approx(1.0)
        assert inputs.trigger_value == pytest.approx(1.0)

    def test_negative_analog_values(self):
        """Test with negative analog values (valid for thumbstick)."""
        inputs = ControllerInputState(
            primary_click=False,
            secondary_click=False,
            thumbstick_click=False,
            menu_click=False,
            thumbstick_x=-1.0,
            thumbstick_y=-1.0,
            squeeze_value=0.0,
            trigger_value=0.0,
        )

        assert inputs.thumbstick_x == pytest.approx(-1.0)
        assert inputs.thumbstick_y == pytest.approx(-1.0)

    def test_invalid_pose(self):
        """Test controller pose with is_valid=False."""
        position = Point(1.0, 2.0, 3.0)
        orientation = Quaternion(0.0, 0.0, 0.0, 1.0)
        pose = Pose(position, orientation)
        controller_pose = ControllerPose(pose, False)

        assert controller_pose.is_valid is False
        # Pose data is still present even if not valid
        assert controller_pose.pose.position.x == pytest.approx(1.0)


class TestControllerSnapshotRecordTimestamp:
    """Tests for ControllerSnapshotRecord with DeviceDataTimestamp."""

    def test_construction_with_timestamp(self):
        """Test ControllerSnapshotRecord carries DeviceDataTimestamp."""
        grip = ControllerPose(
            Pose(Point(1.0, 0.0, 0.0), Quaternion(0.0, 0.0, 0.0, 1.0)), True
        )
        aim = ControllerPose(
            Pose(Point(0.0, 1.0, 0.0), Quaternion(0.0, 0.0, 0.0, 1.0)), True
        )
        inputs = ControllerInputState(True, False, False, False, 0.5, 0.0, 0.8, 1.0)
        data = ControllerSnapshot(grip, aim, inputs)
        ts = DeviceDataTimestamp(1000000000, 2000000000, 3000000000)
        record = ControllerSnapshotRecord(data, ts)

        assert record.timestamp.available_time_local_common_clock == 1000000000
        assert record.timestamp.sample_time_local_common_clock == 2000000000
        assert record.timestamp.sample_time_raw_device_clock == 3000000000
        assert record.data is not None
        assert record.data.inputs.primary_click is True

    def test_default_construction(self):
        """Test default ControllerSnapshotRecord has no data."""
        record = ControllerSnapshotRecord()
        assert record.data is None

    def test_timestamp_fields(self):
        """Test all three DeviceDataTimestamp fields are accessible."""
        data = ControllerSnapshot()
        ts = DeviceDataTimestamp(111, 222, 333)
        record = ControllerSnapshotRecord(data, ts)

        assert record.timestamp.available_time_local_common_clock == 111
        assert record.timestamp.sample_time_local_common_clock == 222
        assert record.timestamp.sample_time_raw_device_clock == 333

File: src/core/schema_tests/cpp/test_controller.cpp
// SPDX-FileCopyrightText: Copyright (c) 2025-2026 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
// SPDX-License-Identifier: Apache-2.0

// Unit tests for the generated Controller FlatBuffer messages.

#include <catch2/catch_approx.hpp>
#include <catch2/catch_test_macros.hpp>
#include <flatbuffers/flatbuffers.h>

// Include generated FlatBuffer headers.
#include <schema/controller_generated.h>
#include <schema/timestamp_generated.h>

#include <type_traits>

// =============================================================================
// Compile-time verification of FlatBuffer field IDs.
// These ensure schema field IDs remain stable across changes.
// VT values are computed as: (field_id + 2) * 2.
// =============================================================================
#define VT(field) (field + 2) * 2

// ControllerSnapshot field IDs (table)
static_assert(core::ControllerSnapshot::VT_GRIP_POSE == VT(0));
static_assert(core::ControllerSnapshot::VT_AIM_POSE == VT(1));
static_assert(core::ControllerSnapshot::VT_INPUTS == VT(2));

// ControllerSnapshotRecord field IDs
static_assert(core::ControllerSnapshotRecord::VT_DATA == VT(0));
static_assert(core::ControllerSnapshotRecord::VT_TIMESTAMP == VT(1));

// =============================================================================
// Compile-time verification that helper types are structs (not tables)
// =============================================================================
static_assert(std::is_trivially_copyable_v<core::ControllerInputState>);
static_assert(std::is_trivially_copyable_v<core::ControllerPose>);
static_assert(std::is_trivially_copyable_v<core::DeviceDataTimestamp>);

// =============================================================================
// ControllerInputState Tests (struct)
// =============================================================================
TEST_CASE("ControllerInputState default construction", "[controller][struct]")
{
    core::ControllerInputState inputs{};

    // Default values should be false/zero.
    CHECK(inputs.primary_click() == false);
    CHECK(inputs.secondary_click() == false);
    CHECK(inputs.thumbstick_click() == false);
    CHECK(inputs.menu_click() == false);
    CHECK(inputs.thumbstick_x() == 0.0f);
    CHECK(inputs.thumbstick_y() == 0.0f);
    CHECK(inputs.squeeze_value() == 0.0f);
    CHECK(inputs.trigger_value() == 0.0f);
}

TEST_CASE("ControllerInputState can store button states", "[controller][struct]")
{
    core::ControllerInputState inputs(true, true, true, true, 0.0f, 0.0f, 0.0f, 0.0f);

    CHECK(inputs.primary_click() == true);
    CHECK(inputs.secondary_click() == true);
    CHECK(inputs.thumbstick_click() == true);
    CHECK(inputs.menu_click() == true);
}

TEST_CASE("ControllerInputState can store analog values", "[controller][struct]")
{
    core::ControllerInputState inputs(false, false, false, false, 0.5f, -0.75f, 0.8f, 1.0f);

    CHECK(inputs.thumbstick_x() == Catch::Approx(0.5f));
    CHECK(inputs.thumbstick_y() == Catch::Approx(-0.75f));
    CHECK(inputs.squeeze_value() == Catch::Approx(0.8f));
    CHECK(inputs.trigger_value() == Catch::Approx(1.0f));
}

// =============================================================================
// ControllerPose Tests (struct)
// =============================================================================
TEST_CASE("ControllerPose default construction", "[controller][struct]")
{
    core::ControllerPose pose{};

    CHECK(pose.is_valid() == false);
}

TEST_CASE("ControllerPose can store pose data", "[controller][struct]")
{
    core::Point position(1.0f, 2.0f, 3.0f);
    core::Quaternion orientation(0.0f, 0.0f, 0.0f, 1.0f);
    core::Pose p(position, orientation);
    core::ControllerPose controller_pose(p, true);

    CHECK(controller_pose.is_valid() == true);
    CHECK(controller_pose.pose().position().x() == Catch::Approx(1.0f));
    CHECK(controller_pose.pose().position().y() == Catch::Approx(2.0f));
    CHECK(controller_pose.pose().position().z() == Catch::Approx(3.0f));
}

// =============================================================================
// DeviceDataTimestamp Tests (struct)
// =============================================================================
TEST_CASE("DeviceDataTimestamp default construction", "[controller][struct]")
{
    core::DeviceDataTimestamp timestamp{};

    CHECK(timestamp.available_time_local_common_clock() == 0);
    CHECK(timestamp.sample_time_local_common_clock() == 0);
    CHECK(timestamp.sample_time_raw_device_clock() == 0);
}

TEST_CASE("DeviceDataTimestamp can store timestamp values", "[controller][struct]")
{
    core::DeviceDataTimestamp timestamp(1000000000LL, 2000000000LL, 3000000000LL);

    CHECK(timestamp.available_time_local_common_clock() == 1000000000LL);
    CHECK(timestamp.sample_time_local_common_clock() == 2000000000LL);
    CHECK(timestamp.sample_time_raw_device_clock() == 3000000000LL);
}

// =============================================================================
// ControllerSnapshotT Tests (table native type)
// =============================================================================
TEST_CASE("ControllerSnapshotT can store complete controller state", "[controller][table]")
{
    // Create grip pose
    core::Point grip_pos(1.0f, 2.0f, 3.0f);
    core::Quaternion grip_orient(0.0f, 0.0f, 0.0f, 1.0f);
    core::Pose grip_p(grip_pos, grip_orient);
    core::ControllerPose grip_pose(grip_p, true);

    // Create aim pose
    core::Point aim_pos(4.0f, 5.0f, 6.0f);
    core::Quaternion aim_orient(0.0f, 0.707f, 0.0f, 0.707f);
    core::Pose aim_p(aim_pos, aim_orient);
    core::ControllerPose aim_pose(aim_p, true);

    // Create inputs
    core::ControllerInputState inputs(true, false, true, false, 0.5f, -0.5f, 0.8f, 1.0f);

    core::ControllerSnapshotT snapshot;
    snapshot.grip_pose = std::make_shared<core::ControllerPose>(grip_pose);
    snapshot.aim_pose = std::make_shared<core::ControllerPose>(aim_pose);
    snapshot.inputs = std::make_shared<core::ControllerInputState>(inputs);

    CHECK(snapshot.grip_pose->is_valid() == true);
    CHECK(snapshot.aim_pose->is_valid() == true);
    CHECK(snapshot.inputs->primary_click() == true);
    CHECK(snapshot.inputs->trigger_value() == Catch::Approx(1.0f));
}

// =============================================================================
// ControllerSnapshotRecordT Tests (table native type)
// =============================================================================
TEST_CASE("ControllerSnapshotRecordT default construction", "[controller][native]")
{
    core::ControllerSnapshotRecordT record;

    CHECK(record.data == nullptr);
}

// =============================================================================
// ControllerSnapshotRecord Serialization Tests
// =============================================================================
TEST_CASE("ControllerSnapshotRecord serialization and deserialization", "[controller][serialize]")
{
    flatbuffers::FlatBufferBuilder builder;

    // Create struct fields
    core::Point pos(1.0f, 2.0f, 3.0f);
    core::Quaternion orient(0.0f, 0.0f, 0.0f, 1.0f);
    core::Pose p(pos, orient);
    core::ControllerPose grip(p, true);
    core::ControllerInputState inputs(true, false, false, false, 0.5f, 0.0f, 0.5f, 0.5f);

    auto snapshot_offset = core::CreateControllerSnapshot(builder, &grip, &grip, &inputs);
    core::ControllerSnapshotRecordBuilder record_builder(builder);
    record_builder.add_data(snapshot_offset);
    builder.Finish(record_builder.Finish());

    // Deserialize
    auto* deserialized = flatbuffers::GetRoot<core::ControllerSnapshotRecord>(builder.GetBufferPointer());

    CHECK(deserialized->data()->inputs()->primary_click() == true);
}

TEST_CASE("ControllerSnapshotRecord can be unpacked from buffer", "[controller][serialize]")
{
    flatbuffers::FlatBufferBuilder builder;

    // Build an empty ControllerSnapshot table
    auto snapshot_offset = core::CreateControllerSnapshot(builder);
    core::ControllerSnapshotRecordBuilder record_builder(builder);
    record_builder.add_data(snapshot_offset);
    builder.Finish(record_builder.Finish());

    // Deserialize to table
    auto* table = flatbuffers::GetRoot<core::ControllerSnapshotRecord>(builder.GetBufferPointer());

    // Unpack to native
    auto unpacked = std::make_unique<core::ControllerSnapshotRecordT>();
    table->UnPackTo(unpacked.get());

    CHECK(unpacked->data != nullptr);
}

// =============================================================================
// ControllerSnapshotRecord Tests (timestamp lives on the Record wrapper)
// =============================================================================
TEST_CASE("ControllerSnapshotRecord serialization with DeviceDataTimestamp", "[controller][serialize]")
{
    flatbuffers::FlatBufferBuilder builder;

    auto record = std::make_shared<core::ControllerSnapshotRecordT>();
    record->data = std::make_shared<core::ControllerSnapshotT>();

    core::Point pos(1.0f, 2.0f, 3.0f);
    core::Quaternion orient(0.0f, 0.0f, 0.0f, 1.0f);
    core::Pose p(pos, orient);
    core::ControllerPose grip(p, true);
    core::ControllerInputState inputs(true, false, false, false, 0.5f, 0.0f, 0.5f, 0.5f);

    record->data->grip_pose = std::make_shared<core::ControllerPose>(grip);
    record->data->aim_pose = std::make_shared<core::ControllerPose>(grip);
    record->data->inputs = std::make_shared<core::ControllerInputState>(inputs);
    record->timestamp = std::make_shared<core::DeviceDataTimestamp>(1000000000LL, 2000000000LL, 3000000000LL);

    auto offset = core::ControllerSnapshotRecord::Pack(builder, record.get());
    builder.Finish(offset);

    auto deserialized = flatbuffers::GetRoot<core::ControllerSnapshotRecord>(builder.GetBufferPointer());

    CHECK(deserialized->timestamp()->available_time_local_common_clock() == 1000000000LL);
    CHECK(deserialized->timestamp()->sample_time_local_common_clock() == 2000000000LL);
    CHECK(deserialized->timestamp()->sample_time_raw_device_clock() == 3000000000LL);
    CHECK(deserialized->data()->inputs()->primary_click() == true);
    CHECK(deserialized->data()->grip_pose()->is_valid() == true);
}

TEST_CASE("ControllerSnapshotRecord can be unpacked with DeviceDataTimestamp", "[controller][serialize]")
{
    flatbuffers::FlatBufferBuilder builder;

    auto original = std::make_shared<core::ControllerSnapshotRecordT>();
    original->data = std::make_shared<core::ControllerSnapshotT>();
    original->timestamp = std::make_shared<core::DeviceDataTimestamp>(111LL, 222LL, 333LL);

    auto offset = core::ControllerSnapshotRecord::Pack(builder, original.get());
    builder.Finish(offset);

    auto fb = flatbuffers::GetRoot<core::ControllerSnapshotRecord>(builder.GetBufferPointer());
    auto unpacked = std::make_shared<core::ControllerSnapshotRecordT>();
    fb->UnPackTo(unpacked.get());

    CHECK(unpacked->timestamp->available_time_local_common_clock() == 111LL);
    CHECK(unpacked->timestamp->sample_time_local_common_clock() == 222LL);
    CHECK(unpacked->timestamp->sample_time_raw_device_clock() == 333LL);
    CHECK(unpacked->data != nullptr);
}
